sys/ztimer/periodic.c

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (C) 2020 Kaspar Schleiser <kaspar@schleiser.de>
+ *               2020 Inria
+ *               2020 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     sys_ztimer_periodic
+ * @{
+ *
+ * @file
+ * @brief       ztimer periodic timer implementation
+ *
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * @}
+ */
+#include <assert.h>
+#include <errno.h>
+#include <stdio.h>
+
+#include "ztimer.h"
+#include "ztimer/periodic.h"
+
+static void _ztimer_periodic_reset(ztimer_periodic_t *timer, ztimer_now_t now)
+{
+    ztimer_now_t target = timer->last + timer->interval;
+    ztimer_now_t offset = target - now;
+
```

Comment:
I think some explicit handling for a missed period make sense. One of the following would make sense to me:

- Run the period as fast as possible, try to get back to the original period by setting `timer->last` to the missed deadline
- Run the period as fast as possible, but set `timer->last` to now
- Set the timer to `timer->last + N * timer->interval`, with `N` being the smallest number for which the result is in the future. (Thus, just giving up on all missed deadlines and try to at least reach any future deadline.)

I'm personally fine with any of the approaches above. But it should be documented, so that users know what they have to expect under too high load.

Reply:
:+1:

any preference? I think I'd personally prefer dropping missed deadlines and schedule the next according to original period.
I hope we don't need to make this configurable...

Reply:
> I hope we don't need to make this configurable...

Me too!

In many soft real time applications (and I do hope that application developers make sure that missed deadlines only occur for soft real time tasks) doing something to late is just as bad as not doing it at all. And this has the advantage that it reduced the load on the system, so that it might recover faster. I guess with those arguments just dropping missed deadlines is a good choice.

Reply:
> I guess with those arguments just dropping missed deadlines is a good choice.

I agree on that, but dropping a missed deadline while at the same time keeping the original period would mean that the ISR would have to loop on adding the interval to the "last" value. For short intervals but long delays, that might take a while, within ISR.

So I now opted for the simplest solution: if the deadline was missed, fire right away.
That way, there's an extra callback per missed tick, but the ISR's are not blocked (or better, are unblocked between callbacks).

Let's see if this works?